src/java.base/share/classes/sun/nio/ch/IOUtil.java

```diff
@@ -474,15 +475,16 @@ static long read(FileDescriptor fd, ByteBuffer[] bufs, int offset, int length,
 
     private static final JavaNioAccess NIO_ACCESS = SharedSecrets.getJavaNioAccess();
 
-    static Runnable acquireScope(ByteBuffer bb, boolean async) {
-        return NIO_ACCESS.acquireSession(bb, async);
+    static MemorySessionImpl acquireScope(ByteBuffer bb, boolean async) {
+        if (async && NIO_ACCESS.isThreadConfined(bb)) {
+            throw new IllegalStateException("Confined session not supported");
```

Comment:
This was the old behaviour which was retained in this PR.

Reply:
I believe `UnsupportedOperationException` would be better for this.

Reply:
> I believe `UnsupportedOperationException` would be better for this.

It would but Per is right that this it is throwing ISE today, it's just that the PR adds an explicit check and that highlights that the wrong exception is thrown. Buffers that are views over memory that is thread confined can't be used with the AsynchronousXXXX APIs and this will needed to be specified. There is further work required in this area too because IOUtil was intended for synchronous I/O. The "async" support was added on a temporary basis and will need to be re-visited at some point.